### ISSUE TITLE
Use of addJetCollection with data in recent CMSSW versions

### ISSUE THREAD

**Issue by kpedro88** [cms-sw/cmssw#20057]
I noticed a ProductNotFound exception when using `addJetCollection()` with data in 92X:
```
----- Begin SkipEvent Exception 03-Aug-2017 17:34:39 CDT-----------------------
An exception of category 'ProductNotFound' occurred while
   [0] Processing  Event run: 275782 lumi: 197 event: 368045582 stream: 0
   [1] Running path 'run'
   [2] Prefetching for module PATJetProducer/'patJetsAK4PFCLEAN'
   [3] Calling method for module MCMatcher/'patJetPartonMatchAK4PFCLEAN'
Exception Message:
Principal::getByToken: Found zero products matching all criteria
Looking for type: std::vector<reco::GenParticle>
Looking for module label: prunedGenParticles
Looking for productInstanceName: 

----- End SkipEvent Exception -------------------------------------------------
```

This occurs despite the producer marking the `patJetPartonMatch` product as a `mayConsume` and the parameter `addGenPartonMatch` being set to false (so the producer never attempts to get the product). The exact same config works fine in 80X. Here are the configs (only difference is a couple of Task-related lines):
`CMSSW_8_0_28`: [test_jetTools_80X.py](http://kjplanet.com/pr/test_jetTools_80X.py)
`CMSSW_9_2_7_patch1`: [test_jetTools_92X.py](http://kjplanet.com/pr/test_jetTools_92X.py)

After some discussion with @Dr15Jones (many thanks!), I learned that prefetching of products was added in 81X. This means that products marked as `mayConsume` are prefetched even if the associated `getByToken()` is never actually called.

I checked a prompt data workflow (where `PATJetProducer` should be called to make miniAOD) and found that the following trick is used:
```
    genJetMatch = cms.InputTag(""),
    genPartonMatch = cms.InputTag(""),
```
(since no product is specified, nothing can be prefetched.)

Therefore, something needs to be updated to handle this correctly for data. I see two possibilities:
1. modify [jetTools.py](https://github.com/cms-sw/cmssw/blob/master/PhysicsTools/PatAlgos/python/tools/jetTools.py) to propagate this tag change when running on data.
2. modify [PATJetProducer](https://github.com/cms-sw/cmssw/blob/CMSSW_9_2_X/PhysicsTools/PatAlgos/plugins/PATJetProducer.cc) to avoid `mayConsume` and instead use conditional `consumes` (only consume the product if the relevant bool is set to true in the config). I think this would make the `iEvent.isRealData()` check in the `produce()` function somewhat pointless, but unfortunately I don't think that information is available early enough to act as a safeguard for the consumes call.

Please backport any fix to 92X, which is currently being used for prompt data analysis.

**Comment by kpedro88** [comment]
@slava77 should this be assigned to reco or analysis?

**Comment by slava77** [comment]
On 8/7/17 9:48 AM, Kevin Pedro wrote:
> @slava77 <https://github.com/slava77> should this be assigned to reco or
> analysis?

can be both, although the initial development of this functionality was
in the analysis

**Comment by Dr15Jones** [comment]
assign reconstruction

**Comment by slava77** [comment]
@zdemirag @gouskos @rappoccio @ahinzmann 
please take a note.

@kpedro88 
are you aware of any follow up on this issue that was not mentioned in this github thread?

**Comment by kpedro88** [comment]
@slava77 not aware of anything

**Comment by ahinzmann** [comment]
here is the fix: #20533
